### PR TITLE
Image pruning fixes

### PR DESCRIPTION
**Pull request by legionus**

### CODE REVIEW

File: pkg/image/prune/prune.go [hunk 1]
```diff
@@ -208,7 +210,7 @@ var _ Pruner = &pruner{}
 //
 // Also automatically remove any image layer that is no longer referenced by any
 // images.
-func NewPruner(options PrunerOptions) Pruner {
+func NewPruner(options PrunerOptions) (Pruner, error) {
```
Comment (mfojtik) [inline pkg/image/prune/prune.go]: godoc

File: pkg/image/prune/prune.go [hunk 2]
```diff
@@ -685,57 +701,76 @@ func pruneStreams(g graph.Graph, imageNodes []*imagegraph.ImageNode, streamPrune
...
-						}
-					}
-					if len(newHistory.Items) == 0 {
-						glog.V(4).Infof("Removing tag %q from status.tags of ImageStream %s", tag, getName(stream))
-						delete(stream.Status.Tags, tag)
+				for i, tagEvent := range history.Items {
+					glog.V(4).Infof("Checking tag event %d with image %q", i, tagEvent.Image)
+
+					if _, ok := imageNodes[tagEvent.Image]; !ok {
+						glog.V(4).Infof("Tag event doesn't match deleted image - keeping")
```
Comment (dmage) [inline pkg/image/prune/prune.go]: New code will produce MUCH better logs, but I still don't like them :-) They are not greppable.

```
glog.V(4).Infof("Checking image stream tag %s:%s: event %d with image %q - keeping, doesn't match deleted image", streamName, tag, i, tagEvent.Image)
glog.V(4).Infof("Checking image stream tag %s:%s: event %d with image %q - keeping, within the allowed range of revisions", streamName, tag, i, tagEvent.Image)
glog.V(4).Infof("Checking image stream tag %s:%s: event %d with image %q - removing, matches deleted image", streamName, tag, i, tagEvent.Image)
```

**Comment by legionus** [comment]
@mfojtik @dmage comments addressed

File: pkg/image/prune/prune.go [hunk 3]
```diff
@@ -766,27 +788,31 @@ func (p *pruner) Prune(
...
 	errs = append(errs, pruneBlobs(p.g, p.registryClient, p.registryURL, prunableComponents, blobPruner)...)
 	errs = append(errs, pruneManifests(p.g, p.registryClient, p.registryURL, prunableImageNodes, manifestPruner)...)
 
 	if len(errs) > 0 {
-		// If we had any errors removing image references from image streams or deleting
-		// layers, blobs, or manifest data from the registry, stop here and don't
-		// delete any images. This way, you can rerun prune and retry things that failed.
+		// If we had any errors deleting layers, blobs, or manifest data from the registry,
+		// stop here and don't delete any images. This way, you can rerun prune and retry
+		// things that failed.
```
Comment (bparees) [inline pkg/image/prune/prune.go]: we need to get smarter about this and still prune the images for which we successfully pruned all the layers/blobs/manifests.

File: pkg/oc/admin/prune/images.go [hunk 4]
```diff
@@ -379,27 +382,34 @@ func (p *describingImageStreamDeleter) GetImageStream(stream *imageapi.ImageStre
...
+	updatedStream, err := p.delegate.UpdateImageStream(stream)
 	if err != nil {
-		fmt.Fprintf(os.Stderr, "error updating image stream %s/%s to remove references to image %s: %v\n", stream.Namespace, stream.Name, image.Name, err)
+		fmt.Fprintf(os.Stderr, "error updating image stream %s/%s to remove image references: %v\n", stream.Namespace, stream.Name, err)
 	}
 
 	return updatedStream, err
 }
 
+func (p *describingImageStreamDeleter) NotifyImageStreamPrune(stream *imageapi.ImageStream, updatedTags []string, deletedTags []string) {
```
Comment (bparees) [inline pkg/oc/admin/prune/images.go]: godoc

**Comment by bparees** [comment]
@legionus is this tied to a bug/issue?

File: pkg/image/prune/prune.go [hunk 5]
```diff
@@ -75,7 +75,9 @@ type ImageStreamDeleter interface {
 	GetImageStream(stream *imageapi.ImageStream) (*imageapi.ImageStream, error)
 	// UpdateImageStream removes all references to the image from the image
 	// stream's status.tags. The updated image stream is returned.
-	UpdateImageStream(stream *imageapi.ImageStream, image *imageapi.Image, updatedTags []string) (*imageapi.ImageStream, error)
+	UpdateImageStream(stream *imageapi.ImageStream) (*imageapi.ImageStream, error)
+	// NotifyImageStreamPrune shows notification about image stream was updated.
```
Comment (miminar) [inline pkg/image/prune/prune.go]: about updated image stream.

File: pkg/image/prune/prune.go [hunk 6]
```diff
@@ -685,57 +696,68 @@ func pruneStreams(g graph.Graph, imageNodes []*imagegraph.ImageNode, streamPrune
...
+					stream.Status.Tags[tag] = newHistory
 				}
-				return err
-			}); err != nil {
-				errs = append(errs, fmt.Errorf("error removing image %s from stream %s: %v",
-					imageNode.Image.Name, getName(streamNode.ImageStream), err))
-				continue
 			}
+
+			updatedStream, err := streamPruner.UpdateImageStream(stream)
```
Comment (bparees) [inline pkg/image/prune/prune.go]: we should know if we actually changed the stream or not, can't we be smarter here?

File: test/extended/images/helper.go [hunk 7]
```diff
@@ -775,6 +775,13 @@ func (c *CleanUpContainer) Run() {
 		}
 	}
 
+	out, err := c.OC.SetNamespace(metav1.NamespaceDefault).AsAdmin().
```
Comment (miminar) [inline test/extended/images/helper.go]: Please add a commentary why is this necessary.

Comment (miminar) [reply test/extended/images/helper.go]: And a TODO to replace this with removals of individual blobs used in the test case.

File: pkg/image/prune/prune.go (hunk 6 above)
Comment (legionus) [reply pkg/image/prune/prune.go]: Yep. Thanks!

Comment (dmage) [reply pkg/image/prune/prune.go]: Though, on the previous step we've selected all image stream that should be modified. If the stream is not changed here, it means someone else concurrently removes objects from the image stream.

File: test/extended/images/helper.go (hunk 7 above)
Comment (legionus) [reply test/extended/images/helper.go]: @miminar Done

File: pkg/image/prune/prune.go [hunk 8]
```diff
@@ -692,57 +696,68 @@ func pruneStreams(g graph.Graph, imageNodes []*imagegraph.ImageNode, streamPrune
...
-						} else {
-							glog.V(4).Infof("Tag event matches deleted image - removing reference")
-							updatedTags.Insert(tag)
-						}
-					}
-					if len(newHistory.Items) == 0 {
-						glog.V(4).Infof("Removing tag %q from status.tags of ImageStream %s", tag, getName(stream))
-						delete(stream.Status.Tags, tag)
+					if _, ok := imageNodes[tagEvent.Image]; !ok {
+						glog.V(4).Infof("Image stream tag %s:%s revision %d - keeping because image %q is not deleted", streamName, tag, i, tagEvent.Image)
```
[the remark this answers is not in the document]
Comment (legionus) [reply pkg/image/prune/prune.go]: @miminar Done

File: pkg/image/prune/prune.go [hunk 9]
```diff
@@ -621,16 +622,19 @@ func imageIsPrunable(g graph.Graph, imageNode *imagegraph.ImageNode, algorithm p
 
 // calculatePrunableImages returns the list of prunable images and a
 // graph.NodeSet containing the image node IDs.
-func calculatePrunableImages(g graph.Graph, imageNodes []*imagegraph.ImageNode, algorithm pruneAlgorithm) ([]*imagegraph.ImageNode, graph.NodeSet) {
-	prunable := []*imagegraph.ImageNode{}
+func calculatePrunableImages(
+	g graph.Graph,
+	imageNodes []*imagegraph.ImageNode,
+	algorithm pruneAlgorithm) (map[string]*imagegraph.ImageNode, graph.NodeSet) {
```
[the remark this answers is not in the document]
Comment (legionus) [reply pkg/image/prune/prune.go]: @miminar Done

File: pkg/image/prune/prune.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (legionus) [reply pkg/image/prune/prune.go]: @miminar I'm a moron. No, we can't do it.

File: pkg/image/prune/prune.go [hunk 10]
```diff
@@ -527,29 +532,31 @@ func addBuildStrategyImageReferencesToGraph(g graph.Graph, strategy buildapi.Bui
...
-			return
+			return err
 		}
 		imageID = id
 	case "DockerImage":
 		ref, err := imageapi.ParseDockerImageReference(from.Name)
 		if err != nil {
 			glog.V(2).Infof("Error parsing DockerImage name %q: %v - skipping", from.Name, err)
-			return
+			return err
```
[the remark this answers is not in the document]
Comment (legionus) [reply pkg/image/prune/prune.go]: @miminar You can not create such strategy:
```
# buildconfigs "cakephp-ex" was not valid:
# * spec.strategy.sourceStrategy.from.name: Invalid value: "foo:bar:baz": not a valid Docker pull specification: invalid reference format
```
Also even if I add such a option, then it will need to specify this option constantly until broken strategy is removed.

File: pkg/oc/admin/prune/images.go (hunk 4 above)
Comment (legionus) [reply pkg/oc/admin/prune/images.go]: @bparees done

**Comment by legionus** [comment]
> is this tied to a bug/issue?

@bparees No.

File: pkg/image/prune/prune.go [hunk 11]
```diff
@@ -671,71 +684,95 @@ func calculatePrunableImageComponents(g graph.Graph) []*imagegraph.ImageComponen
...
 
-				for tag, history := range stream.Status.Tags {
-					glog.V(4).Infof("Checking tag %q", tag)
+	for streamName, streamNode := range prunableStreams {
+		err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+			stream, err := streamPruner.GetImageStream(streamNode.ImageStream)
+			if err != nil {
+				// During prune someone removed imagestream
+				if kerrapi.IsNotFound(err) {
+					return nil
```
Comment (miminar) [inline pkg/image/prune/prune.go]: *nit*: I think it's worth of 4th info level at least.

File: pkg/image/prune/prune.go [hunk 12]
```diff
@@ -671,71 +684,95 @@ func calculatePrunableImageComponents(g graph.Graph) []*imagegraph.ImageComponen
...
+				return nil
+			}
+
+			updatedStream, err := streamPruner.UpdateImageStream(stream)
+			if err == nil {
+				streamPruner.NotifyImageStreamPrune(stream, updatedTags.List(), deletedTags.List())
+				streamNode.ImageStream = updatedStream
+			}
+
+			return err
```
Comment (miminar) [inline pkg/image/prune/prune.go]: *paranoid mode on*: another `NotFound` could occur here as well :*paranoid mode off*

File: pkg/image/prune/prune.go [hunk 13]
```diff
@@ -671,71 +684,95 @@ func calculatePrunableImageComponents(g graph.Graph) []*imagegraph.ImageComponen
...
-					}
-					if len(newHistory.Items) == 0 {
-						glog.V(4).Infof("Removing tag %q from status.tags of ImageStream %s", tag, getName(stream))
-						delete(stream.Status.Tags, tag)
+					if _, ok := prunableImageNodes[tagEvent.Image]; !ok {
+						glog.V(4).Infof("Image stream tag %s:%s revision %d - keeping because image %q is not deleted", streamName, tag, i, tagEvent.Image)
+						newHistory.Items = append(newHistory.Items, tagEvent)
 					} else {
-						stream.Status.Tags[tag] = newHistory
+						glog.V(4).Infof("Image stream tag %s:%s revision %d - removing because image %q matches deleted image", streamName, tag, i, tagEvent.Image)
```
Comment (miminar) [inline pkg/image/prune/prune.go]: This branch applies also to **not** prunable images.

File: pkg/image/prune/prune.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (miminar) [reply pkg/image/prune/prune.go]: One of the prior docker/distribution rebases made parsing of references stricter. I can easily imagine really old build strategy with a relaxed reference value. I can also imagine this parser becoming even stricter.

> Also even if I add such a option, then it will need to specify this option constantly until broken strategy is removed.

If the error is printed, the admin will have the opportunity to act and tune/delete the object. Make sure to include namespace/name of the build config.

File: pkg/image/prune/prune.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (miminar) [reply pkg/image/prune/prune.go]: > No, we can't do it.

Can you elaborate?

File: pkg/image/prune/prune.go (hunk 11 above)
Comment (legionus) [reply pkg/image/prune/prune.go]: @miminar Done

File: pkg/image/prune/prune.go (hunk 12 above)
Comment (legionus) [reply pkg/image/prune/prune.go]: @miminar Done

File: pkg/image/prune/prune.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (legionus) [reply pkg/image/prune/prune.go]: @miminar What you suggested cannot be implemented in this place because if the user adds a new image, we'll delete as non-existent.

File: pkg/image/prune/prune.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (legionus) [reply pkg/image/prune/prune.go]: @miminar Done

File: pkg/image/prune/prune.go (hunk 5 above)
Comment (legionus) [reply pkg/image/prune/prune.go]: @miminar Done

File: pkg/image/prune/prune.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (legionus) [reply pkg/image/prune/prune.go]: @miminar I mean that we have to be tolerant to unknown images.

File: pkg/image/prune/prune.go [hunk 14]
```diff
@@ -526,38 +539,39 @@ func addBuildStrategyImageReferencesToGraph(g graph.Graph, strategy buildapi.Bui
...
-			return
+			return fmt.Errorf("error parsing ImageStreamImage name %q: %v", from.Name, err)
 		}
 		imageID = id
 	case "DockerImage":
 		ref, err := imageapi.ParseDockerImageReference(from.Name)
 		if err != nil {
-			glog.V(2).Infof("Error parsing DockerImage name %q: %v - skipping", from.Name, err)
-			return
+			return fmt.Errorf("error parsing DockerImage name %q: %v", from.Name, err)
```
Comment (bparees) [inline pkg/image/prune/prune.go]: I thought I saw agreement to make this a warning, not fail the entire prune process?  That would be my inclination.

Same for the ImageStreamImage check above.

If the reference is invalid, it clearly is not going to reference an image and keep us from pruning that image.

Comment (legionus) [reply pkg/image/prune/prune.go]: > I thought I saw agreement to make this a warning, not fail the entire prune process? That would be my inclination.

@bparees Yes, but my answer was that you can't create such reference.

> If the reference is invalid, it clearly is not going to reference an image and keep us from pruning that image.

You can not be sure that this Strategy is not going to reference an image. You just don't know. I believe that in the conditions when we are not sure we need to save user data. In the end the user will see the error and can decide to remove this build/buildconfig or fix reference.

https://github.com/openshift/origin/pull/16717#discussion_r143468046

Comment (bparees) [reply pkg/image/prune/prune.go]: > @bparees Yes, but my answer was that you can't create such reference.

I think relying on that guarantee is a mistake.  Bugs in the buildconfig validation could allow the creation of such a reference and then pruning is broken for no good reason.

> You can not be sure that this Strategy is not going to reference an image.

I don't care what it may do in the future.  I know it's not currently referencing an image.  What the user may update it to reference in the future when they fix their typo is irrelevant.  It may have tried to reference an image, but it is not referencing an image (it has an invalid reference, so it's not referencing anything).

> I believe that in the conditions when we are not sure we need to save user data.

What are we saving?  Can you give me a scenario where a user manages to create an image that we'd prune here despite an invalid reference existing in the system, that would ultimately be a problem?

If your argument is "the user meant to reference something but they did it wrong so we have to save everything until we know what they meant to reference", well what if they meant to reference "ruby" and they typed it as "riby"?  It's a "valid" reference, so it won't trip this error, but the "ruby" image is going to get cleaned up in that case too.  There's only so much protection we can offer.

> In the end the user will see the error and can decide to remove this build/buildconfig or fix reference.

The admin running this is not the same as the user who created the broken reference.  So your solution requires the admin to either go edit the user's object themselves, or track down the user and discuss how to resolve it.  Neither of those are ideal situations for an admin who's just trying to free up space (not to mention that this is frequently run as an automated job that may or may not have good monitoring in place).

In your referenced comment thread you said "done" but it's not obvious to me what that means.  You didn't add the suggested "force" flag and you didn't make this a warning.  (And to be clear, I prefer the latter).

Comment (legionus) [reply pkg/image/prune/prune.go]: > I think relying on that guarantee is a mistake. Bugs in the buildconfig validation could allow the creation of such a reference and then pruning is broken for no good reason.

@bparees I'll change it, but it's your decision.

> In your referenced comment thread you said "done" but it's not obvious to me what that means

My `done` was about:

> If the error is printed, the admin will have the opportunity to act and tune/delete the object. Make sure to include namespace/name of the build config.

and I added the ability to see the name of the object.

**Comment by legionus** [comment]
@bparees Also, I should note that `pruneStreams` has an unsolved bug: If you specify the `--keep-tag-revisions=0` option, then this code can remove the tag completely, but the prune process does not count tag references process does not take into account the references to the tags and it can break Builds/BuildConfigs if they have references to these tags.

File: pkg/image/prune/prune.go (hunk 14 above)
Comment (bparees) [reply pkg/image/prune/prune.go]: >  I'll change it, but it's your decision.

I asked a pretty direct question: 

What are we saving? Can you give me a scenario where a user manages to create an image that we'd prune here despite an invalid reference existing in the system, that would ultimately be a problem?

You seem very concerned about something here, but I haven't grasped what it is.  Can you help me understand the scenario you are envisioning where proceeding with pruning is a mistake and why it is a mistake?

I am trying to make an *educated* decision....  so far I am not seeing the risk in proceeding with pruning in this situation and i'd like you to help me be confident that I am not overlooking something.

**Comment by bparees** [comment]
> @bparees Also, I should note that pruneStreams has an unsolved bug: If you specify the --keep-tag-revisions=0 option, then this code can remove the tag completely

Hm.  I suppose you're getting what you asked for if you do that but I agree it's likely to surprise people.

File: pkg/image/prune/prune.go (hunk 14 above)
Comment (legionus) [reply pkg/image/prune/prune.go]: > I asked a pretty direct question:

@bparees At first I did not understand correctly. Sorry, of course I will try to explain my point of view.

> What are we saving?

Since this operation can not be canceled (we send requests to the registry to delete blobs/manifests), then we need to be very careful with the calculation of references. Otherwise, we will permanently delete objects that the user did not want to lose.

If we have references that we can not parse, it means that there is something wrong in the system (incomplete upgrade or something else). In this case, it is not the best time to perform actions that can not be rolled back.

Ignoring "broken" Build/BuildConfig will lead us to the image pruning if there are no other references in the cluster.

> Can you give me a scenario where a user manages to create an image that we'd prune here despite an invalid reference existing in the system, that would ultimately be a problem?

Imagine the situation when in the new version of origin we updated the `github.com/openshift/origin/pkg/image/reference` in such a way (it can be corner cases) that 10% of the old references stopped parsing. If you ignore the reference parsing error, you will lose 10% of the links in the graph. This will lead to a loss of actual data.

Comment (bparees) [reply pkg/image/prune/prune.go]: > Imagine the situation when in the new version of origin we updated the github.com/openshift/origin/pkg/image/reference in such a way (it can be corner cases) that 10% of the old references stopped parsing. If you ignore the reference parsing error, you will lose 10% of the links in the graph. This will lead to a loss of actual data.

Thanks, that is a compelling argument.

Do we at least have enough information to add the namespace containing the bad buildconfig reference?  Right now i'm not convinced we're printing enough information to help an admin go find the bad reference and fix it.  They essentially would have to dump all the buildconfigs in all the namespaces and then grep for the string.  We should do more to help point them directly to where the problem is, in these cases.

**Comment by legionus** [comment]
>  I suppose you're getting what you asked for.

@bparees Obviously, the user did not ask us to turn the cluster into a non-consistent state. From our [documentation](https://docs.openshift.org/latest/admin_guide/pruning_resources.html#pruning-images):

> In order to prune images that are no longer required by the system due to age, status, or exceed limits, administrators may run the following command:
> `$ oadm prune images [<options>]`

and
> 1. Specify --keep-tag-revisions=0 with the oadm prune images command.
>
> This action will effectively remove all the tags from all the namespaces with underlying images, unless they are younger or they are referenced by objects younger than the specified threshold.

So we do not have to delete something that is still required by the system.

File: pkg/image/prune/prune.go (hunk 14 above)
Comment (legionus) [reply pkg/image/prune/prune.go]: > Do we at least have enough information to add the namespace containing the bad buildconfig reference? 

@bparees Right now, I output an error to the user and specifying `<namespace>/<name>` of the Build/BuildConfig. Is this information enough ?

**Comment by bparees** [comment]
> @bparees Obviously, the user did not ask us to turn the cluster into a non-consistent state.

but they did ask us to remove all the tags from the system, so long as the associated images are older than a certain age.

> So we do not have to delete something that is still required by the system.

sorry, I can't follow this statement.  What behavior are you proposing would be correct for "--keep-tag-revisions=0"?

File: pkg/image/prune/prune.go (hunk 14 above)
Comment (bparees) [reply pkg/image/prune/prune.go]: yes, i hadn't seen that you were catching this error and wrapping it w/ additional information.  I see it now.

**Comment by legionus** [comment]
> sorry, I can't follow this statement. What behavior are you proposing would be correct for "--keep-tag-revisions=0"?

@bparees Let's consider the situation:

We have two Images:
* Image `sha256:image1` was created month ago;
* Image `sha256:image2` was created 5 minutes ago;

ImageStream `myproject/foo` (was created month ago) has two tags:
* Tag `v1` that points to Image `sha256:image1`;
* Tag `v2` that points to Image `sha256:image2`;

One running Pod that points to the Image `sha256:image2`.
One Build that points to ImageStreamTag `myproject/foo:v1`.

Now let's run the command:
`oc adm prune images --keep-tag-revisions=3 --keep-younger-than=60m`

We will make the graph and add the dependences:
* Pod will give us dependency to `sha256:image2` ([here is the code](https://github.com/openshift/origin/blob/master/pkg/image/prune/prune.go#L428))
* ImageStream will not give us  strong references at all [here is the code](https://github.com/openshift/origin/blob/master/pkg/image/prune/prune.go#L294)
* Build will not give us  strong references too because it points to ImageStreamTag [here is the code](https://github.com/openshift/origin/blob/master/pkg/image/prune/prune.go#L525-L542).

So according to the graph, we can remove the Image `sha256:image1` because there is no reference to it and it is older than necessary. And since it was the only one in the tag, we remove the `v1` tag from ImageStream `myproject/foo`.

Now we have a Build that points to `myproject/foo:v1` tag that no longer exists.

This happened because we did not count the reference to the `v1` tag itself when it was deleted. If we did this, we would not be able to remove it and therefore we could not remove the last image from this tag.

**Comment by bparees** [comment]
> Build will not give us strong references too because it points to ImageStreamTag

hm.  It seems like for object that have an imagestreamtag value, we should be resolving what image that imagestreamtag currently points to, and counting that as a hard reference from the object(buildconfig in this case) to the image.  Is there a reason we did not do that?

**Comment by legionus** [comment]
> Is there a reason we did not do that?

@bparees The tag may change during pruning. Since Build points to the tag we must protect this tag from pruning. Otherwise, if the tag changes, we will remove it anyway. Imagine the situation that after the start of the pruning process, someone updated the tag to something that we want to remove.

`<offtopic>` In the current conditions, the pruning can not be done correctly. There will always be the possibility of a race condition.`</offtopic>`

Moreover, this process with a certain probability can damage the registry storage (if you use S3). I mean, if you're pushing data at the same time, there's a chance to get broken images.

**Comment by bparees** [comment]
btw you have the same problem with a DeploymentConfig that is not currently deployed, I would imagine.  You have no hard reference to an image, just a reference from the DeploymentConfig to an ImageStreamTag.

Which again makes me think we should be resolving the reference to an image and treating that as a strong reference between the image and the deploymentconfig(or buildconfig).

If someone updates the reference it mid-pruning you're still not any worse off than you are today where you ignore the reference entirely.

**Comment by legionus** [comment]
> so do we protect it or not?

@bparees No. You can see it [here](https://github.com/legionus/origin/blob/eeb73882d4f0c72d35ea7d93eeddb6ea5c05f812/pkg/image/prune/prune.go#L748-L754).

> As you say, we're never going to solve 100% of potential race conditions since we don't have a transactional lock on the graph.

We can change the conditions and make it possible. As a crazy idea: Mover the pruning process to the master and block (not reject) write requests while the pruning is in progress. In this case, we guarantee the immutability of the state. Also, we can move blobs/manifests pruning to the registry (we already have hard pruning), rather than sending millions of delete requests, which are rather slow.

> I think (if only for my education) we need to have a level set conversation on what the pruning algorithm is and why. 

@bparees I did not know anything about image pruning process a weeks ago )))
So I'm in the same position as you.

> btw you have the same problem with a DeploymentConfig that is not currently deployed, I would imagine. You have no hard reference to an image, just a reference from the DeploymentConfig to an ImageStreamTag.

Sure. I do not know of all the places where there can be such tag references, so I listed only builds and buildconfigs. But of course there are certainly more of them.

> If someone updates the reference it mid-pruning you're still not any worse off than you are today where you ignore the reference entirely.

Sure. It's better to take into account these references than to ignore them completely.

Honestly, I do not understand why the graph is used to prune images. There is enough map of images with a reference counter. This will work well only if we are sure that the state of the cluster can not change, but since we ignore this fact, we can not take it into account.
